src/main/java/org/spongepowered/api/event/entity/living/trader/VillagerEvent.java

```diff
@@ -0,0 +1,106 @@
+package org.spongepowered.api.event.entity.living.trader;
+
+import org.spongepowered.api.data.type.Profession;
+import org.spongepowered.api.entity.Entity;
+import org.spongepowered.api.entity.living.Hostile;
+import org.spongepowered.api.entity.living.golem.IronGolem;
+import org.spongepowered.api.entity.living.monster.raider.Raider;
+import org.spongepowered.api.entity.living.monster.zombie.ZombieEntity;
+import org.spongepowered.api.entity.living.trader.Villager;
+import org.spongepowered.api.event.Cancellable;
+import org.spongepowered.api.event.Event;
+import org.spongepowered.api.raid.Raid;
+import org.spongepowered.api.util.annotation.eventgen.GenerateFactoryMethod;
+
+/**
+ * An event which involves a {@link Villager}.
+ */
+public interface VillagerEvent extends Event {
+
+	/**
+	 * Gets the {@link Villager} involved with this event.\
+	 *
+	 * @return The villager.
+	 */
+	Villager getVillager();
+
+	/**
+	 * Fired when a {@link Villager}'s profession changes.
+	 *
+	 * <p>This can include both gaining an losing a {@link Profession}.</p>
+	 */
+	@GenerateFactoryMethod
+	interface ProfessionChange extends VillagerEvent, Cancellable {
+
+		/**
+		 * Gets the {@link Profession} the villager will change to.
+		 *
+		 * @return The {@link Villager}'s next {@link Profession}.
+		 */
+		Profession getNextProfession();
+
+		/**
+		 * Sets the {@link Villager}'s next {@link Profession}.
+		 */
+		void setNextProfession(Profession profession);
+	}
+
+	/**
+	 * Fired when a {@link Villager} levels up it's {@link Profession}.
+	 */
+	@GenerateFactoryMethod
+	interface ProfessionLevelUp extends VillagerEvent, Cancellable {
+
+		/**
+		 * Gets the {@link Villager}'s current {@link Profession} level.
+		 *
+		 * @return The {@link Profession} level.
+		 */
+		default int getCurrentProfessionLevel() {
+			this.getVillager().professionLevel().get();
```

Comment:
This needs to be passed to the event factory in the impl, never default and point to the source like this.

Reply:
Right, and especially since a plugin can change the profession level value between event listeners.

Reply:
Mind you that this would technically be possible to be part of `ChangeValueEvent` as a sub class `LevelUpProfession` that is based on the `Keys.PROFESSION_LEVEL`. I'll be making adjustments to `event-impl-gen` to allow annotating `Key`s fields to generate an event based on it.